ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2838,292 @@ public sealed class Bot : IAsyncDisposable {
+
+				if (result.Count > 0) {
+					await Actions.SendInventory(result).ConfigureAwait(false);
+				}
+			} finally {
+				SendCompleteSetsSemaphore.Release();
+			}
+		}
+
+		private async Task<Dictionary<uint, byte>?> LoadCardsPerSet(HashSet<uint> appIDs) {
+			if (appIDs == null) {
+				throw new ArgumentNullException(nameof(appIDs));
+			}
+
+			switch (ASF.GlobalConfig?.OptimizationMode) {
+				case GlobalConfig.EOptimizationMode.MinMemoryUsage:
+					Dictionary<uint, byte> result = new Dictionary<uint, byte>(appIDs.Count);
+
+					foreach (uint appID in appIDs) {
+						byte cardCount = await ArchiWebHandler.GetCardCountForGame(appID).ConfigureAwait(false);
+
+						if (cardCount == 0) {
+							return null;
+						}
+
+						result.Add(appID, cardCount);
+					}
+
+					return result;
+				default:
+					IEnumerable<Task<(uint AppID, byte Cards)>> tasks = appIDs.Select(async appID => (AppID: appID, Cards: await ArchiWebHandler.GetCardCountForGame(appID).ConfigureAwait(false)));
+					IList<(uint AppID, byte Cards)> results = await Utilities.InParallel(tasks).ConfigureAwait(false);
+
+					if (results.Select(tuple => tuple.Cards).Any(cards => cards == 0)) {
+						return null;
+					}
+
+					return results.ToDictionary(res => res.AppID, res => res.Cards);
+			}
+		}
+
+		internal static HashSet<Steam.Asset> GetItemsForFullBadge(IReadOnlyCollection<Steam.Asset> inventory, Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), (uint SetsToExtract, byte CardsPerSet)> amountsToExtract) {
+			if ((inventory == null) || (amountsToExtract == null) || (amountsToExtract.Count == 0) || (inventory.Count == 0)) {
+				throw new ArgumentNullException($"{nameof(inventory)} || {nameof(amountsToExtract)}");
+			}
+
+			HashSet<Steam.Asset> result = new HashSet<Steam.Asset>();
+			Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), Dictionary<ulong, List<Steam.Asset>>> itemsPerClassIDPerSet = inventory.GroupBy(item => (item.RealAppID, item.Type, item.Rarity)).ToDictionary(grouping => grouping.Key, grouping => grouping.GroupBy(item => item.ClassID).ToDictionary(group => group.Key, group => group.ToList()));
+
+			foreach (((uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity) set, (uint setsToExtract, byte cardsPerSet)) in amountsToExtract) {
+				if (!itemsPerClassIDPerSet.TryGetValue(set, out Dictionary<ulong, List<Steam.Asset>>? itemsPerClassID)) {
+					continue;
+				}
+
+				if (cardsPerSet != itemsPerClassID.Count) {
+					continue;
+				}
+
+				foreach (List<Steam.Asset> itemsOfClass in itemsPerClassID.Values) {
+					long classRemaining = setsToExtract;
+					int i = 0;
+
+					while ((classRemaining > 0) && (i < itemsOfClass.Count)) {
+						Steam.Asset item = itemsOfClass[i++];
+
+						if (classRemaining - item.Amount < 0) {
+							Steam.Asset itemToSend = item.CreateShallowCopy();
+							itemToSend.Amount = (uint) classRemaining;
```

Comment:
Previously it was `Convert.ToUInt16`, which converted to `ushort`, can you confirm that you meant converting to `uint`?

Reply:
That was a mistake. The correct one is `uint`. The usage of `uint` is also the reason `classRemaining` is `long` in the first place.